### PR TITLE
ZK-4524: blocking all user keyboard input during processing message showing

### PR DESCRIPTION
**Pull request by b952426**
ZK-4524: blocking all user keyboard input during processing message showing

### CODE REVIEW

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zk/src/archive/web/js/zk/au.js [hunk 1]
```diff
@@ -1505,6 +1506,7 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
 					message: msg
 				});
 			});
+			jq(w).on('keydown', zk.$void);
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/au.js:1509]: Use two anchors around w, and send tab (or shift+tab) to the next focusable element

File: zk/src/archive/web/js/zk/au.js [hunk 2]
```diff
@@ -1513,6 +1515,7 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
 	/** Removes the busy message covering the whole browser.
 	 */
 	clearBusy: function (uuid) {
+		var w = uuid ? Widget.$(uuid) : null;
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/au.js:1518]: move to if

File: zk/src/archive/web/js/zk/utl.js [hunk 3]
```diff
@@ -488,6 +489,7 @@ zUtl.parseMap("a='b c',c=de", ',', "'\"");
 				e._applyMask.destroy();
 				e._applyMask = null;
 			}
+		jq('html').off('keydown', zk.$void);
```
Comment (scribetw) [inline zk/src/archive/web/js/zk/utl.js:492]: consider moving to z_mask instead

File: zktest/src/archive/test2/F96-ZK-4524.zul [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	<button label="processMask" onClick="Thread.sleep(3000);">
+	</button>
+	<button label="box show busy" onClick='Clients.showBusy(target,"processing")'/>
+	<button label="box clear busy" onClick='Clients.clearBusy(target)'/>
+	<button label="show busy" onClick='Clients.showBusy("processing");'/>
+	<div id="target" style="border: 1px solid blue;background: #cde" height="100px">
+		This box will be covered by the busy message
+		<textbox></textbox>
+	</div>
+	<textbox></textbox>
```
Comment (scribetw) [inline zktest/src/archive/test2/F96-ZK-4524.zul:26]: `<textbox/>`

File: zktest/src/archive/test2/F96-ZK-4524.zul [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+Copyright (C) 2021 Potix Corporation. All Rights Reserved.
+-->
+<zk>
+	<label multiline="true">
+	</label>
+	<button label="processMask" onClick="Thread.sleep(3000);">
+	</button>
+	<button label="box show busy" onClick='Clients.showBusy(target,"processing")'/>
+	<button label="box clear busy" onClick='Clients.clearBusy(target)'/>
+	<button label="show busy" onClick='Clients.showBusy("processing");'/>
```
Comment (scribetw) [inline zktest/src/archive/test2/F96-ZK-4524.zul:21]: add a `<textbox />` here

File: zktest/test/java/org/zkoss/zktest/zats/test2/F96_ZK_4524Test.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		actions.sendKeys("component showBusy").perform();
+		Assert.assertEquals("", textbox1.val());
+		Assert.assertEquals("component showBusy", textbox2.val());
+		click(jq("@button:eq(2)"));
+		waitResponse();
+		actions.sendKeys(Keys.TAB, Keys.TAB, "component showBusy").perform();
+		Assert.assertEquals("component showBusy", textbox1.val());
+		click(jq("@button:eq(3)"));
+		waitResponse();
+		actions.sendKeys(Keys.TAB, "browser showBusy", Keys.TAB, Keys.TAB, "browser showBusy").perform();
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F96_ZK_4524Test.java:54]: extract full mask case

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zk/src/archive/web/js/zk/au.js [hunk 7]
```diff
@@ -1535,6 +1539,19 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
 				}
 			}
+			jq('html').off('keydown', zk.$void);
+		}
+	},
+	/** Disable specified component keydown while showBusy(Component, String) is called.
+	 * while component receive shiftKey and tabKey, focus previous input element.
+	 * while component receive tabKey, focus next input element.
+	 */
+	disableKeydown: function (e) {
```
Comment (b952426) [inline zk/src/archive/web/js/zk/au.js:1549]: move to other suitable place

File: zk/src/archive/web/js/zk/au.js [hunk 8]
```diff
@@ -1535,6 +1539,19 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+		e.preventDefault();
+		if (e.which == 9 && e.shiftKey) { //tabKey and shiftKey
+			jq(this).prev(':input, [href], [tabindex]:not([tabindex="-1"])').focus();
+		} else if (e.which == 9) { //tabKey
```
Comment (b952426) [inline zk/src/archive/web/js/zk/au.js:1553]: e.originalEvent.key==zKeys.TAB

File: zk/src/archive/web/js/zk/au.js [hunk 9]
```diff
@@ -1535,6 +1539,19 @@ zAu.cmd0 = /*prototype*/ { //no uuid at all
...
+			jq(this).next(':input, [href], [tabindex]:not([tabindex="-1"])').focus();
```
Comment (b952426) [inline zk/src/archive/web/js/zk/au.js:1554]: jq(this).prevAll('button, [href], input, select, textarea,[tabindex]').filter(':not([tabindex="-1"])').first()

File: zk/src/archive/web/js/zk/au.js [hunk 10]
```diff
@@ -1145,6 +1145,42 @@ zAu.beforeSend = function (uri, req, dt) {
...
+	skipBusyWidget: function (e) {
+		e.preventDefault();
+		var focusableElements = '[href], :input,[tabindex]',
+			isTabKey = e.key == zKeys.TAB,
+			elem;
+		if (isTabKey && e.shiftKey) {
+			elem = zAu._getFirstFocusableElement(jq(this).prevAll(focusableElements));
+		} else if (isTabKey) {
+			elem = zAu._getFirstFocusableElement(jq(this).nextAll(focusableElements));
+		}
```
Comment (b952426) [inline zk/src/archive/web/js/zk/au.js:1161]: //if widget prev/next focusable element is null, focus last/first focusable element.
only tab. shifTab preventDefault

**Review by scribetw — APPROVED** [review]

**Review by plijyfes — APPROVED** [review]

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**